server/src/main/java/org/elasticsearch/index/mapper/BooleanFieldMapper.java

```diff
@@ -237,23 +277,53 @@ protected void parseCreateField(ParseContext context, List<IndexableField> field
             return;
         }
 
-        Boolean value = context.parseExternalValue(Boolean.class);
-        if (value == null) {
-            XContentParser.Token token = context.parser().currentToken();
-            if (token == XContentParser.Token.VALUE_NULL) {
-                if (fieldType().nullValue() != null) {
-                    value = fieldType().nullValue();
+        Boolean value = null;
+        try {
+            value = context.parseExternalValue(Boolean.class);
+            if (value == null) {
+                final XContentParser parser = context.parser();
+                XContentParser.Token token = parser.currentToken();
+                if (token == XContentParser.Token.VALUE_NULL) {
+                    if (fieldType().nullValue() != null) {
+                        value = fieldType().nullValue();
+                    }
+                } else {
+                    if (indexCreatedVersion.onOrAfter(Version.V_6_0_0_alpha1)) {
+                        if(!parser.isBooleanValue()) {
+                            // need to consume array/object, otherwise parser doesnt skip to the next field as expected
+                            parser.skipChildren();
+
+                            throw new IllegalArgumentException(parser.currentToken().toString());
```

Comment:
Is this the reason for the try/catch that was introduced? If so, can the control flow somehow be done withouth thrrowing the exception here?

Reply:
see my longer comment slightly below. if the legacy stuff is removed this will be simplified and you will need to double check again.

Reply:
Consider the throw a nasty way to keep flow control while i await your commentary whether i should keep/remove legacy lenient support that was recently removed.